File: src/main/java/cz/cvut/fel/plichjan/distmesh/matlab/Matlab.java
package cz.cvut.fel.plichjan.distmesh.matlab;

import cz.cvut.fel.plichjan.distmesh.inputs.IDistanceFunction;
import delaunay.Pnt;
import delaunay.Triangle;
import delaunay.Triangulation;
import org.apache.log4j.Logger;

import java.util.*;

/**
 */
public class Matlab {
    public static final Logger logger = Logger.getLogger(Matlab.class);

    /** returns the distance from 1.0 to the next largest double-precision number */
    public static final double EPS = Math.nextUp(1.) - 1.;

    public static double[] vector(double first, double step, double last) {
        double[] vector = new double[(int) ((last - first) / step) + 1];
        for (int i = 0; i < vector.length; i++) {
            vector[i] = i * step + first;
        }
        return vector;
    }

    public static int[] vector(int i0, int iN) {
        final int length = iN - i0;
        final int[] ints = new int[length];
        for (int i = 0; i < length; i++) {
            ints[i] = i + i0;
        }
        return ints;
    }

    public static <E> List<E> filter(final List<E> src, ITest<E> test) {
        List<E> dst = new ArrayList<E>(src.size());
        for (int i = 0; i < src.size(); i++) {
            if (test.call(i, src.get(i))) {
                dst.add(src.get(i));
            }
        }
        return dst;
    }

    public static <E> void filterSelf(final List<E> src, ITest<E> test) {
        int i = 0;
        for (Iterator<E> iterator = src.iterator(); iterator.hasNext(); ) {
            E next = iterator.next();
            if (!test.call(i, next)) {
                iterator.remove();
            }
            i++;
        }
    }

    public static <E> List<E> filter(List<E> src, int[] ix) {
        List<E> dst = new ArrayList<E>(ix.length);
        for (int i : ix) {
            dst.add(src.get(i));
        }
        return dst;
    }

    public static List<int[]> delaunayn2(List<Pnt> p, double[][] bbox, final IDistanceFunction fd, final double geps) {
        final Pnt lb = new Pnt(bbox[0]);
        final Pnt tr = new Pnt(bbox[1]);
        final Pnt wh = tr.subtract(lb);
        Triangle tri = new Triangle(lb.subtract(wh), lb.add(new Pnt(wh.coord(0) * 3, 0)), lb.add(new Pnt(0, wh.coord(1) * 3)));
        Triangulation triangulation = new Triangulation(tri);
        for (int i = 0; i < p.size(); i++) {
            Pnt point = p.get(i);
            point.setIndex(i);
            if (fd.call(point.get(0), point.get(1)) < geps) {
                triangulation.delaunayPlace(point);
            }
        }

        ArrayList<int[]> ints = new ArrayList<int[]>(triangulation.size());
        for (Triangle triangle : triangulation) {
            if (Collections.disjoint(tri, triangle)) {
                Pnt p1 = triangle.get(0);
                Pnt p2 = triangle.get(1);
                Pnt p3 = triangle.get(2);
                ints.add(new int[] {p1.getIndex(), p2.getIndex(), p3.getIndex()});
            }
        }
        if (logger.isDebugEnabled()) {
            logger.debug("initEquiTriangles done. " + p.size() + " tr: " + ints.size());
        }
        return ints;
    }

    public static List<Pnt> asPntList(double[][] p) {
        final List<Pnt> list = new ArrayList<Pnt>(p.length);
        for (double[] point : p) {
            list.add(new Pnt(point));
        }
        return list;
    }

    public static double[][] toArray(List<Pnt> p) {
        final double[][] arr = new double[p.size()][];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = p.get(i).getData();
        }
        return arr;
    }

    public static List<Pnt> computeCentroids(List<Pnt> p, List<int[]> t) {
        final List<Pnt> pmid = new ArrayList<Pnt>(t.size());
        for (int[] tr : t) {
            Pnt sum = new Pnt(new double[p.get(0).dimension()]);
            for (int i : tr) {
                sum = sum.add(p.get(i));
            }
            pmid.add(sum.scale(1. / tr.length));
        }
        return pmid;
    }

    public static double[][] computeCentroids(double[][] p, int[][] t) {
        return toArray(computeCentroids(asPntList(p), Arrays.asList(t)));
    }
}

File: src/main/kotlin/cz/cvut/fel/plichjan/distmesh/matlab/Matlab.kt
package cz.cvut.fel.plichjan.distmesh.matlab

import cz.cvut.fel.plichjan.distmesh.inputs.IDistanceFunction
import delaunay.Pnt
import delaunay.Triangle
import delaunay.Triangulation
import org.apache.log4j.Logger
import java.util.*
import kotlin.math.nextUp

/**
 */
object Matlab {
    val logger: Logger = Logger.getLogger(Matlab::class.java)

    /** returns the distance from 1.0 to the next largest double-precision number */
    val EPS: Double = 1.0.nextUp() - 1.0

    @JvmStatic
    fun vector(first: Double, step: Double, last: Double): DoubleArray {
        val size = ((last - first) / step).toInt() + 1
        return DoubleArray(size) { i -> i * step + first }
    }

    @JvmStatic
    fun vector(i0: Int, iN: Int): IntArray {
        val length = iN - i0
        return IntArray(length) { i -> i + i0 }
    }

    @JvmStatic
    fun <E> filter(src: List<E>, test: ITest<E>): List<E> {
        val dst = ArrayList<E>(src.size)
        for (i in src.indices) {
            if (test.call(i, src[i])) {
                dst.add(src[i])
            }
        }
        return dst
    }

    @JvmStatic
    fun <E> filterSelf(src: MutableList<E>, test: ITest<E>) {
        var i = 0
        val iterator = src.iterator()
        while (iterator.hasNext()) {
            val next = iterator.next()
            if (!test.call(i, next)) {
                iterator.remove()
            }
            i++
        }
    }

    @JvmStatic
    fun <E> filter(src: List<E>, ix: IntArray): List<E> {
        val dst = ArrayList<E>(ix.size)
        for (i in ix) {
            dst.add(src[i])
        }
        return dst
    }

    @JvmStatic
    fun delaunayn2(p: List<Pnt>, bbox: Array<DoubleArray>, fd: IDistanceFunction, geps: Double): List<IntArray> {
        val lb = Pnt(*bbox[0])
        val tr = Pnt(*bbox[1])
        val wh = tr.subtract(lb)
        val tri = Triangle(lb.subtract(wh), lb.add(3.0, Pnt(wh.coord(0), 0.0)), lb.add(3.0, Pnt(0.0, wh.coord(1))))
        val triangulation = Triangulation(tri)
        for (i in p.indices) {
            val point = p[i]
            point.index = i
            if (fd.call(point.coord(0), point.coord(1)) < geps) {
                triangulation.delaunayPlace(point)
            }
        }

        val ints = ArrayList<IntArray>(triangulation.size)
        for (triangle in triangulation) {
            // Check if the triangle is disjoint from the initial large triangle
            var overlapsInitial = false
            for (v in tri) {
                if (triangle.contains(v)) {
                    overlapsInitial = true
                    break
                }
            }
            if (!overlapsInitial) {
                val p1 = triangle[0]
                val p2 = triangle[1]
                val p3 = triangle[2]
                ints.add(intArrayOf(p1.index, p2.index, p3.index))
            }
        }
        if (logger.isDebugEnabled) {
            logger.debug("initEquiTriangles done. ${p.size} tr: ${ints.size}")
        }
        return ints
    }

    @JvmStatic
    fun asPntList(p: Array<DoubleArray>): List<Pnt> {
        return p.map { Pnt(*it) }
    }

    @JvmStatic
    fun toArray(p: List<Pnt>): Array<DoubleArray> {
        return Array(p.size) { i -> p[i].getData() }
    }

    @JvmStatic
    fun computeCentroids(p: List<Pnt>, t: List<IntArray>): List<Pnt> {
        val pmid = ArrayList<Pnt>(t.size)
        for (tr in t) {
            var sum = Pnt(*DoubleArray(p[0].dimension()))
            for (i in tr) {
                sum = sum.add(p[i])
            }
            pmid.add(sum.scale(1.0 / tr.size))
        }
        return pmid
    }

    @JvmStatic
    fun computeCentroids(p: Array<DoubleArray>, t: Array<IntArray>): Array<DoubleArray> {
        return toArray(computeCentroids(asPntList(p), t.toList()))
    }
}
